### PR TITLE
Create the lexer

### PR DESCRIPTION
**Pull request by Lyxal**

### CODE REVIEW

**Review by PyGamer0 — COMMENTED** [review]
Why dont you use an Enum for TokenTypes?

File: vyxal/lexer.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+            tokens.append(Token(TokenType.NAME, contextual_token_value))
+        elif head in "k∆øÞ¨":
+            if source:
+                tokens.append(Token(TokenType.GENERAL, head + source.popleft()))
+        else:
+            tokens.append(Token(TokenType.GENERAL, head))
+    return tokens
+
+
+if __name__ == "__main__":
```
Comment (ysthakur) [inline vyxal/lexer.py:181]: Might want to run these as proper tests (compare against an expected output and throw an error if it doesn't work).

Comment (ysthakur) [inline vyxal/lexer.py:181]: I think these should be run as proper tests (compare against an expected output and throw an error if it doesn't work). It doesn't feel right to be running this module all by itself, and IMO the only module with an `if __name__ == "__main__"` should be the interpreter/compiler.

**Review by ysthakur — CHANGES_REQUESTED** [review]
Might want to run these as proper tests (compare against an expected output and throw an error if it doesn't work). It doesn't feel right to be running this module all by itself. Other than that, looks pretty good to me.

I can always remove this request if y'all feel that this is okay, though. Can you indicate that using the thumbs up/down emojis?

**Review by ysthakur — APPROVED** [review]

File: vyxal/lexer.py (hunk 1 above)
Comment (Tams-Tams) [reply vyxal/lexer.py]: What do mean by proper tests?

Comment (ysthakur) [reply vyxal/lexer.py]: Well, each group of tests would be in a function. I was thinking all the tests for the lexer would be in a `lexer_tests.py` file with all the other tests. That way, if something fails, we'll know when we push.

Comment (Tams-Tams) [reply vyxal/lexer.py]: who are you, actually?

Comment (Tams-Tams) [reply vyxal/lexer.py]: A username?

File: vyxal/lexer.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+structures, they need to be turned into tokens representing the
+different components of a program. For the full specification on token
+types, go to documents/specs/Lexer.md
+"""
+
+import collections
+import string
+
+
+class TokenType:
```
Comment (PyGamer0) [inline vyxal/lexer.py:13]: why not an Enum??

**Review by PyGamer0 — COMMENTED** [review]
mmm

File: vyxal/lexer.py (hunk 2 above)
Comment (Tams-Tams) [reply vyxal/lexer.py:13]: Yes yes

File: vyxal/lexer.py (hunk 1 above)
Comment (ysthakur) [reply vyxal/lexer.py]: I don't understand. In real life? I'm not telling you that, for obvious reasons.

**Review by Tams-Tams — APPROVED** [review]

File: vyxal/lexer.py (hunk 2 above)
Comment (Lyxal) [reply vyxal/lexer.py:13]: no. that seems _too_ much

**Review by ysthakur — APPROVED** [review]

**Pull request merged**